File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/javascript/find_first_duplicate.js
function findFirstDuplicate(arr) {
  const uniqueSet = new Set();

  for (const value of arr) {
    if (uniqueSet.has(value)) {
      return value;
    }

    uniqueSet.add(value);
  }

  return -1;
}


if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 3");
  console.log("=>", findFirstDuplicate([2, 1, 3, 3, 2]));

  console.log("");

  console.log("Expecting: -1");
  console.log("=>", findFirstDuplicate([1, 2, 3, 4]));
}

module.exports = findFirstDuplicate;



// Please add your pseudocode to this file
// And a written explanation of your solution

// Initialize an empty set called uniqueSet
//iterate through the array,  if the value is in uniqueSet, return the value
//   else (the value is not already in the uniqueSet) add the value the set array

// return -1 if no duplicate found in array

File: 01-week-1--starter-algorithms/01-day-2--find-first-duplicate/ruby/find_first_duplicate.rb
require 'set'

def find_first_duplicate(arr)
  uniqueSet = Set.new

  arr.each do |value|
    return value if uniqueSet.include?(value)

    uniqueSet.add(value)
  end

  -1
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 3"
  puts "=>", find_first_duplicate([2, 1, 3, 3, 2])

  puts

  puts "Expecting: -1"
  puts "=>", find_first_duplicate([1, 2, 3, 4])

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution

 # init  empty set called uniqueSet
 # 
 # iterate through the array, if value is in uniqueSet, return the value(duplicate)
 # or else, add the value to the set
 # if no duplicate is found, return -1
